File: KotlinBuilderBase.scala
package org.jetbrains.plugins.kotlinConverter

trait KotlinBuilderBase {
  private val builder = new StringBuffer()
  private var i = 0
  private val indentStep = 2
  def str(v: Any): Unit =
    builder.append(v)

  def nl(): Unit = {
    str("\n")
    str(" " * i)
  }

  def indent(): Unit = {
    i += indentStep
    nl()
  }
  def unIndent(): Unit = {
    i -= indentStep
    nl()
  }

  def rep[T](values: Seq[T], sep: => Unit)(h: T => Unit): Unit =
    if (values.nonEmpty) {
      values.init.foreach { value =>
        h(value)
        sep
      }
      h(values.last)
    }

  def rep[T](values: Seq[T], sep: String)(h: T => Unit): Unit =
    rep(values, str(sep))(h)

  def repNl[T](values: Seq[T])(h: T => Unit): Unit =
    rep(values, nl)(h)

  def text: String = builder.toString
}

File: KotlinBuilderBase..kt
package org.jetbrains.plugins.kotlinConverterpublic

abstract class KotlinBuilderBase {
    val builder = StringBuffer()
    var i: Int = 0
    val indentStep = 2
    public fun str(v: Any): Unit = builder.append(v)
    public fun nl(): Unit {
        str("\n")
        str(" ".repeat(i))
    }

    public fun indent(): Unit {
        i += indentStep
        nl()
    }

    public fun unIndent(): Unit {
        i -= indentStep
        nl()
    }

    public fun rep(values: List<T>, sep: Unit, h: (T) -> Unit): Unit = if (values.isNotEmpty()) {
        values.init().forEach { value ->
            h(value)
            sep
        }
        h { values.last(it) }
    }

    public fun rep(values: List<T>, sep: String, h: (T) -> Unit): Unit = rep()(values, str(sep), h)
    public fun repNl(values: List<T>, h: (T) -> Unit): Unit = rep()(values, nl(), h)
    public fun opt(value: T?, h: (T) -> Unit): Unit = value.foreach(h)
    public fun text(): String = builder.toString
}